### ISSUE TITLE
ROI: Folders - Mockups - Measuring Tool - Step 3.1 - Folder and ROI interactions - Part 1

### ISSUE THREAD

**Issue by gusferguson** [openmicroscopy/design#14]
**Note:**
**There is quite a bit of repetition and overlap with Step 3.0 in this and other 3.x steps, but once it is decided what will be done in step 3 these will all be tidied up and rationalised.**

- Newer mockups supersede older.
- Sometimes loose the wood for the trees when doing a long series of mockups, so some details may be missing or wrong. Any of these need commented on please.

**Maintaining Context and Copy vs Move Delete vs Orphaned**

As pointed out by @wmoore - the ROI Folder - ROI add, move, orphan has some of the same issues related to copy/link/delete/unlink in Project > Dataset > Image.

A few points to try and nail this down:

- ROI-Folders has orphaned ROIs cf not really with images
- should still represent the ROI (as orphaned) even if it is in one or more Folders
- ROI have single image context whereas Folders have group wide context 
- potential confusion
- if a Folder doesn’t contain an ROI from that image it will not appear in the Measurement Tool window i.e. would need another window to allow interaction with them
- need to make clear the difference between:
 - deleting ROI - “delete”
 - moving ROI out of Folder - orphaning - “move”
 - moving ROI into folder from orphaned - “move”
 - moving ROI from folder to folder - “move”
 - adding ROI to second or subsequent Folder - “copy and paste”
- need very clear distinction between deleting an ROI and removing it from a Folder

- consider dealing with ROI interaction and Folder interaction completely separately to help with disambiguation 
- this is largely done below

May then have situation where some actions will involve more steps than otherwise necessary but these can be worked on.

*The order of the mockups below is a bit cumbersome - will sort out with tidy-up.*

**Adding selected ROIs or ROI folders**

A dialog based on the existing Add Tab dialog can be used for adding selected ROIs or ROI folders to new or existing ROI folder.

Destination Folder default in dialog:
- is the selected folder if ROIs selected (in Measurement Tool) 
- should be blank if ROI Folder selected.

Selected Folder or Folder that selected ROIs are in:
- greyed out in LH panel of dialog

![image 9 scoping step 3 - roi folder window dest default](https://cloud.githubusercontent.com/assets/1552703/12923435/a759a05a-cf4d-11e5-9bfa-738f1d84df0c.png)

Tooltips in the destination Folder box give hierarchy of selected destination Folder

![image 10 scoping step 3 - roi folder window dest tt 1](https://cloud.githubusercontent.com/assets/1552703/12923443/b3b3ac9c-cf4d-11e5-9ade-fcb3cc4b6e29.png)

A different destination folder can be selected from the drop-down list.
- list would have to contain all ROI folders in the group
- this would include duplicates e.g. duplicate subfolder names in different screens
- tooltip showing hierarchy would disambiguate

![image 11 scoping step 3 - roi folder window dest tt 2](https://cloud.githubusercontent.com/assets/1552703/12923455/bd2fdd86-cf4d-11e5-8a5d-e63aa0423b8b.png)

New Folders can be created using the text fields at the bottom left:

![image 12 scoping step 3 - roi folder new](https://cloud.githubusercontent.com/assets/1552703/12923467/ca3a734c-cf4d-11e5-8275-f8a49ad4f2c0.png)

And added to RH panel with or without existing ROI Folders

![image 13 scoping step 3 - roi folder new added](https://cloud.githubusercontent.com/assets/1552703/12923483/d41dfde8-cf4d-11e5-99c3-1776fd7a67c6.png)

If just orphan ROIs were selected in Management Tool, then the destination folder drop-down would be empty and the destination folders would be selected from the LH panel and added to RH panel using the arrows.

![image 14 scoping step 3 - roi folder existing added](https://cloud.githubusercontent.com/assets/1552703/12923495/dec695d4-cf4d-11e5-86f4-3b8d367f817c.png)

Tooltips and breadcrumbs used in Measurement Tool for context

Selected Folder/ROI - breadcrumb location at bottom of window

![image 15 scoping step 3 - selected](https://cloud.githubusercontent.com/assets/1552703/12923504/e79f421e-cf4d-11e5-8a81-a6ec2922953c.png)

Tooltips (artificially showing tooltips for ROIs and Folders together):

![image 16 scoping step 3 - tooltips](https://cloud.githubusercontent.com/assets/1552703/12923515/ee432f7c-cf4d-11e5-880c-9e66077a6511.png)

Continued in:

 ROI: Folders - Mockups - Measuring Tool - Step 3.1 - Folder and ROI interactions - Part 2

**Comment by gusferguson** [comment]
Still working on it

**Comment by dominikl** [comment]
- I'm not quite sure about the purpose of the "destination folder" box, aren't the destination folders the selected folders? 
- There's still the problem with the tag selection dialog in general with respect to multi selection. Select multiple ROIs which are in different folders (some common, some not common ones), which folders are "selected", is it the intersection or the union? 
- I still assume, we have clearly distinct action for adding and removing ROIs to/from folders, right? I. e. "Add" action: Popup selection dialog, add Rois to selected folders; "Remove" action: Popup selection dialog, but now remove the ROIs from the selected folders. This will be different from the Tag selection, where the same action is used for adding/removing tags (selection (move to right) means adding, de-selection (move to left) means removing), which leads to unresolvable situations for multi-selection (see https://github.com/openmicroscopy/openmicroscopy/pull/4399 ), I'd like to avoid that for the ROI folder selection.

**Comment by mtbc** [comment]
Also need to ensure in QA that destination folders for move do not include the folder's own children.

**Comment by will-moore** [comment]
The "Destination ROI Folder" really confused me. Took me a long time re-reading the notes etc to work out what it was doing. Is it for choosing a Parent folder for any NEW ROI folders that you happen to create? I don't think this should happen in this dialog since it complicates it too much. The dialog should be for a single purpose of putting the selected ROIs / folders into existing (or new) folders. Not to also put new folders into a chosen parent. In fact for Tags, we don't even use the Tag dialog for putting Tags into Tag-Sets, right? So maybe we should adopt the same strategy here?

**Comment by hflynn** [comment]
I was just struggling with the 'Destination ROI Folder' bit too actually. It's not at all clear what it's for in the context of the rest of the panel.

**Comment by gusferguson** [comment]
The problem I was trying to address with the destination drop-down is that:
1. all ROI Folders in the group would need to be shown in the LH panel - it could become very crowded
2. enabling more than one folder to be moved at a time.

Happy to go with simpler and "repeat as needed" that if people don't like the destination selection.

**Comment by mtbc** [comment]
Can I use a filter to narrow down the list of destination folders?

**Comment by gusferguson** [comment]
Yes there is a filter box above the LH panel.

**Comment by mtbc** [comment]
Ah, great, thanks, I thought that might just be about picking the folder to be moved.

**Comment by will-moore** [comment]
I think this is getting a bit more complex than is needed. I think the only things we need are:

 - a button to add folders to the view (you can pick existing folder or create a new one)
 - then you can build your hierarchy by drag and drop (with a right-click Move, Copy-To & Remove options to make this easier).

Both the drag-and-drop and the right-click menu would only allow you to move to ROI folders that are already in the view - NOT all the folders in OMERO because this is too complex. I don't think we need an equivalent to the Tag dialog. What we are doing here is building hierarchies just like for P/D/I tree, and we don't have such a dialog for that.
When you add a folder to the view, it would appear in the hierarchy with all it's parent AND child folders. The "Add existing folder" dialog should probably present all Folders as a tree, so you can see if you're picking a folder with child or parent folders.

**Comment by pwalczysko** [comment]
Sorry, not getting the ``Destination folder`` either.

**Comment by dominikl** [comment]
Yes, drag&drop is definitely the way one wants to organize the folders. I'll investigate next, if/how that's possible to implement.

**Comment by gusferguson** [comment]
All the material in this is now covered in https://github.com/openmicroscopy/design/issues/30 or earlier.
